icefoganalytics/internal-data-portal
# Part 1: User: All Datasets Page

Relates to:
- https://github.com/icefoganalytics/internal-data-portal/issues/7
- See [All Datasets page - /datasets](https://docs.google.com/document/d/1tP63Ob4XPBGlASPHVf1GnNb3hlIkybmWBqJDA61u3J0/edit#heading=h.hyc5ymxijixo)
- partially completes https://github.com/icefoganalytics/internal-data-portal/issues/20

# Context

Out of Scope:
- add create, update, delete for data owner
- search

## All Datasets page - /datasets

### Back-end routes

* /datasets/search?term=... - magic search endpoint
    * Filter by tags/keywords
    * Filter by department
    * filter by ownership
* /datasets - generic paginated data sets list, not sortable at this time.


### Components

* `BaseLayout.vue` - see previous
* `DatasetSearch.vue` - Generates a search query against all datasets and redirects the user to a page with the appropriate fields filled in?
* `DatasetTable.vue` - list all datasets
    * Dataset - Dataset#name
    * Description - Dataset#description
    * Keywords - Tags through Taggings
    * Owner - Dataset#owner -> department
    * Access - magic aggregate field that describes the type of access the current user has available against the database, should be serialized in the back-end from the AccessGrants for the dataset.
    * (unlabeled) - aggregate of possible actions and states depending on available AccessGrants and current AccessRequests, and the type of access request (i.e. subscriptions). This is pretty complex logic and would benefit from testing.
        * `RequestAccessButton.vue`
            * Either opens a request access dialog or redirects to a request access form
        * `SubscribeToDatasetButton.vue`
            * Either opens a subscription request dialog or redirects to a subscription request form
        * Subscribed state
        * Approved state
        * Awaiting approval state

**Additional context**

![image](https://github.com/icefoganalytics/internal-data-portal/assets/23045206/2da7a524-3171-4fc3-a6b6-610619201428)

# Implementation

Add a bunch of models, around half of all models have not been implemented.
Add some QA Scenarios to easily generate fake data for testing.
Add vue-i18n for consistent translation of enums in the front-end.

# Screenshots

http://localhost:8080/datasets
> Note that Access and action have not been implemented.
![image](https://github.com/icefoganalytics/internal-data-portal/assets/23045206/805570b7-4c7c-464a-9a96-bdabe1959e71)

# Testing Instructions

1. Run the test suite via `dev test` (or `dev test_api`)
2. Boot the app via `dev up`
3. Log in to the app at http://localhost:8080
4. From the http://localhost:8080/dashboard page, create a dataset if you haven't already.
5. Go back to the dashboard and trigger the "link random tags" scenario.
6. Go to the "View Datasets" page via the button.
7. Check that the dataset name, descriptions, keywords and owner information are as you expect.
> Note that Access and action have not been implemented do to massive complexity requiring implementation of a back-end test suite.
